usr/src/uts/common/fs/zfs/zfs_vfsops.c

```diff
@@ -848,6 +848,148 @@ zfs_owner_overquota(zfsvfs_t *zfsvfs, znode_t *zp, boolean_t isgroup)
 	return (zfs_fuid_overquota(zfsvfs, isgroup, fuid));
 }
 
+static zfsvfs_kstats_t empty_zfsvfs_kstats = {
+	{ "writes",	KSTAT_DATA_UINT64 },
+	{ "nwritten",	KSTAT_DATA_UINT64 },
+	{ "reads",	KSTAT_DATA_UINT64 },
+	{ "nread",	KSTAT_DATA_UINT64 },
+};
+
+static int
+zfsvfs_kstats_update(kstat_t *ksp, int rw)
+{
+	zfsvfs_t *zfsvfs = ksp->ks_private;
+	zfsvfs_kstats_t *zs = ksp->ks_data;
+
+	ASSERT3P(zfsvfs->z_iokstat, ==, ksp);
+
+	if (rw == KSTAT_WRITE)
+		return (EACCES);
+
+	zs->zk_writes.value.ui64 =
+	    aggsum_value(&zfsvfs->z_aggsum_stats.zas_writes);
+	zs->zk_nwritten.value.ui64 =
+	    aggsum_value(&zfsvfs->z_aggsum_stats.zas_nwritten);
+	zs->zk_reads.value.ui64 =
+	    aggsum_value(&zfsvfs->z_aggsum_stats.zas_reads);
+	zs->zk_nread.value.ui64 =
+	    aggsum_value(&zfsvfs->z_aggsum_stats.zas_nread);
+
+	return (0);
+}
+
+static void
+zfsvfs_kstats_create(zfsvfs_t *zfsvfs)
+{
+	ASSERT3P(zfsvfs->z_iokstat, ==, NULL);
+
+	/*
+	 * There should not be anything wrong with having kstats for
+	 * snapshots. Since we are not sure how useful they would be
+	 * though nor how much their memory overhead would matter in
+	 * a filesystem with many snapshots, we skip them for now.
+	 */
+	if (zfsvfs->z_issnap)
+		return;
+
+	/*
+	 * We are limited by KSTAT_STRLEN for the kstat's name here
+	 * which is a lot less that the string length of a dataset's
+	 * path. Thus, we distinguish handles by their objset IDs
+	 * prepended by the pool's load_guid. Note that both numbers
+	 * are formatted in hex. See breakdown in comment below.
+	 *
+	 * We use the pool's load_guid because it is guaranteed to
```

Comment:
Do you mean that the pool-guid can be changed without re-import of the corresponding pool, that will require recreate of kstat-structures?

Reply:
Thank you for looking at the review @Ramzec.

Yes. The pool-guid can change without re-import (e.g. we do a `zpool reguid <pool>`).
Thus, if we were using the pool-guild as part of the kstat name, if the pool's guid changed, all these kstats would be left off with the old guid.

Those kstats are created when a dataset is mounted and destroyed when the dataset is unmounted. So it makes more sense to use the pool's load_guid as this would keep things consistent.

Given the above, recreating all these structures every time the guid changes would be very complicated and unnecessary. Does this make sense?

Reply:
Yep, that what I supposed. Would be nice if you put that into the comment that will make it more clear.